common/persistence/nosql/nosqlplugin/cassandra/gocql/session.go

```diff
@@ -0,0 +1,172 @@
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package gocql
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/gocql/gocql"
+
+	"go.temporal.io/server/common/resolver"
+	"go.temporal.io/server/common/service/config"
+)
+
+var _ Session = (*session)(nil)
+
+const (
+	sessionRefreshMinInternal = 5 * time.Second
+)
+
+type (
+	session struct {
+		config       config.Cassandra
+		resolver     resolver.ServiceResolver
+		atomic.Value // *gocql.Session
+
+		sync.Mutex
+		sessionInitTime time.Time
+	}
+)
+
+func NewSession(
+	config config.Cassandra,
+	resolver resolver.ServiceResolver,
+) (*session, error) {
+
+	gocqlSession, err := initSession(config, resolver)
+	if err != nil {
+		return nil, err
+	}
+
+	session := &session{
+		config:          config,
+		sessionInitTime: time.Now().UTC(),
+	}
+	session.Value.Store(gocqlSession)
+	return session, nil
+}
+
+func (s *session) refresh() error {
+	s.Lock()
+	defer s.Unlock()
+
+	if time.Now().UTC().Sub(s.sessionInitTime) < sessionRefreshMinInternal {
+		return nil
+	}
+
+	gocqlSession, err := initSession(s.config, s.resolver)
+	if err != nil {
+		return err
+	}
+
+	s.sessionInitTime = time.Now().UTC()
+	s.Value.Store(gocqlSession)
+	return nil
+}
+
+func initSession(
+	config config.Cassandra,
+	resolver resolver.ServiceResolver,
+) (*gocql.Session, error) {
+	cluster, err := NewCassandraCluster(config, resolver)
+	if err != nil {
+		return nil, err
+	}
+	cluster.ProtoVersion = 4
+	cluster.Consistency = config.Consistency.GetConsistency()
+	cluster.SerialConsistency = config.Consistency.GetSerialConsistency()
+	cluster.Timeout = 10 * time.Second
+	return cluster.CreateSession()
+}
+
+func (s *session) Query(
+	stmt string,
+	values ...interface{},
+) Query {
+	q := s.Value.Load().(*gocql.Session).Query(stmt, values...)
+	if q == nil {
+		return nil
+	}
+	return &query{
+		session: s,
+		query:   q,
+	}
+}
+
+func (s *session) NewBatch(
+	batchType BatchType,
+) Batch {
+	b := s.Value.Load().(*gocql.Session).NewBatch(mustConvertBatchType(batchType))
+	if b == nil {
+		return nil
+	}
+	return &batch{
+		session: s,
+		batch:   b,
+	}
+}
+
+func (s *session) ExecuteBatch(
+	b Batch,
+) (retError error) {
+	defer func() { s.handleError(retError) }()
+
+	return s.Value.Load().(*gocql.Session).ExecuteBatch(b.(*batch).batch)
+}
+
+func (s *session) MapExecuteBatchCAS(
+	b Batch,
+	previous map[string]interface{},
+) (_ bool, _ Iter, retError error) {
+	defer func() { s.handleError(retError) }()
+
+	applied, iter, err := s.Value.Load().(*gocql.Session).MapExecuteBatchCAS(b.(*batch).batch, previous)
+	if iter == nil {
+		return applied, nil, err
+	}
+	return applied, iter, err
+}
+
+func (s *session) Close() {
+	session, ok := s.Value.Load().(*gocql.Session)
```

Comment:
can this race with refresh() and cause refresh to create a new session after we closed the existing one?

Reply:
yes, there can be race causing session not to be closed.